File: ios/Packages/Primitives/Sources/TransferDataExtra.swift
// Copyright (c). Gem Wallet. All rights reserved.

import BigInt
import Foundation

public struct TransferDataExtra: Equatable, Sendable {
    public let to: String
    public let gasLimit: BigInt?
    public let gasPrice: GasPriceType?
    public let data: Data?
    public let outputType: TransferDataOutputType
    public let outputAction: TransferDataOutputAction
    public let transactionType: TransactionType
    public let approval: ApprovalData?

    public init(
        to: String,
        gasLimit: BigInt? = .none,
        gasPrice: GasPriceType? = .none,
        data: Data? = .none,
        outputType: TransferDataOutputType = .encodedTransaction,
        outputAction: TransferDataOutputAction = .send,
        transactionType: TransactionType = .smartContractCall,
        approval: ApprovalData? = nil,
    ) {
        self.to = to
        self.gasLimit = gasLimit
        self.gasPrice = gasPrice
        self.data = data
        self.outputType = outputType
        self.outputAction = outputAction
        self.transactionType = transactionType
        self.approval = approval
    }
}

extension TransferDataExtra: Hashable {}

File: core/crates/primitives/src/transfer_data_extra.rs
use num_bigint::BigInt;
use serde::{Deserialize, Serialize};

use crate::swap::ApprovalData;
use crate::{GasPriceType, TransactionType, TransferDataOutputAction, TransferDataOutputType};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransferDataExtra {
    pub to: String,
    pub gas_limit: Option<BigInt>,
    pub gas_price: Option<GasPriceType>,
    pub data: Option<Vec<u8>>,
    pub output_type: TransferDataOutputType,
    pub output_action: TransferDataOutputAction,
    pub transaction_type: TransactionType,
    pub approval: Option<ApprovalData>,
}

impl TransferDataExtra {
    pub fn data_as_str(&self) -> Result<&str, &'static str> {
        let bytes = self.data.as_ref().ok_or("missing data")?;
        std::str::from_utf8(bytes).map_err(|_| "data is not valid utf8")
    }
}

impl Default for TransferDataExtra {
    fn default() -> Self {
        Self {
            to: "".to_string(),
            gas_limit: None,
            gas_price: None,
            data: None,
            output_type: TransferDataOutputType::EncodedTransaction,
            output_action: TransferDataOutputAction::Send,
            transaction_type: TransactionType::SmartContractCall,
            approval: None,
        }
    }
}
